lib/src/Pos/Lrc/DB/Leaders.hs

```diff
@@ -44,13 +62,49 @@ prepareLrcLeaders ::
        , HasConfiguration
        )
     => m ()
-prepareLrcLeaders =
-    whenNothingM_ (getLeaders 0) $
-        putLeaders 0 genesisLeaders
+prepareLrcLeaders = do
+    whenNothingM_ (getLeadersForEpoch 0) $
+        putLeadersForEpoch 0 genesisLeaders
+    -- Smooth migration for CSE-240.
+    initLeaders 1
+  where
+    initLeaders :: MonadDB m => EpochIndex -> m ()
+    initLeaders i = do
+        whenNothingM_ (getLeader $ SlotId i (UnsafeLocalSlotIndex 0)) $ do
+            maybeLeaders <- getLeadersForEpoch i
+            case maybeLeaders of
+                Just leaders -> do
+                    putLeadersForEpochSeparately i leaders
+                    initLeaders (i + 1)
+                Nothing -> pure ()
 
 ----------------------------------------------------------------------------
 -- Keys
 ----------------------------------------------------------------------------
 
-leadersKey :: EpochIndex -> ByteString
-leadersKey = mappend "l/" . serialize'
+leadersForEpochKey :: EpochIndex -> ByteString
+leadersForEpochKey = mappend "l/" . serialize'
+
+leaderKey :: HasProtocolConstants => SlotId -> ByteString
+leaderKey = mappend "ls/" . serialize' . flattenSlotId
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+putLeadersForEpochAllAtOnce :: MonadDB m => EpochIndex -> SlotLeaders -> m ()
+putLeadersForEpochAllAtOnce epoch = putBi (leadersForEpochKey epoch)
+
+putLeadersForEpochSeparately :: MonadDB m => EpochIndex -> SlotLeaders -> m ()
+putLeadersForEpochSeparately epoch leaders =
+    putBatchBi putLeadersForEachSlotOps
+  where
+    putLeadersForEachSlotOps :: [(ByteString, StakeholderId)]
+    putLeadersForEachSlotOps = [
+            (leaderKey $ mkSlotId epoch i, leader)
+            | (i, leader) <- zip [1..] $ toList leaders
+        ]
+    mkSlotId :: EpochIndex -> Word16 -> SlotId
+    mkSlotId epoch' slot =
+        -- Using @UnsafeLocalSlotIndex@ because we trust the callers.
+        SlotId epoch' (UnsafeLocalSlotIndex slot)
```

Comment:
There is `unsafeMkLocalSlotIndex` function which is a bit better because it will `error` if `slot` is out of bounds.